### PR TITLE
[DoDeca] IGC Task

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
This PR mainly adds the [Image Grounded Conversations (IGC)](https://www.aclweb.org/anthology/I17-1047/) task to ParlAI; IGC is one of the 12 [dodecaDialogue](https://parl.ai/projects/dodecadialogue/) tasks.

The IGC task must be manually downloaded from a url and placed in the correct folder; this teacher will automatically download the available images.

I additionally made some changes to the `AbstractImageTeacher`

**Testing steps**
Added `tests/tasks/test_igc.py`. Should only be run locally.

**Logs**
```
$ python tests/tasks/test_igc.py
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216, got 192
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216, got 192
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216, got 192
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
..
----------------------------------------------------------------------
Ran 2 tests in 23.807s

OK
```

**Data tests (if applicable)**
```
$ python parlai/scripts/verify_data.py -t igc
...
274s elapsed: {'exs': 4353, '%done': '100.00%', 'time_left': '0s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 4353, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
```

### CODE REVIEW

File: parlai/tasks/igc/agents.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        """
+
+        dt = opt['datatype'].split(':')[0]
+        dt_str = 'test' if dt == 'test' else 'val'
+        dp = os.path.join(self.get_data_path(opt), f'IGC_crowd_{dt_str}.csv')
+        if not os.path.exists(dp):
+            raise RuntimeError(
+                'Please download the IGC Dataset from '
+                'https://www.microsoft.com/en-us/download/details.aspx?id=55324. '
+                'Then, make sure to put the two .csv files in ParlAI/data/igc/'
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: can you sub in here the the actual datapath? the user might supply --datapath

File: parlai/tasks/igc/agents.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+            'text': text,
+            'image_id': image_id,
+            'episode_done': episode_done,
+            'image': image_features,
+            'labels': labels,
+        }
+
+        return action
+
+    def load_data(self, data_path: str, opt: Dict[str, Any]) -> List[Dict[str, Any]]:
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: opt is an object, `parlai.core.opt.Opt`

Comment (stephenroller) [reply parlai/tasks/igc/agents.py]: thanks for typing :D

File: parlai/tasks/igc/agents.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+class IGCTeacher(AbstractImageTeacher):
+    """
+    Teacher for IGC Task.
+
+    See https://arxiv.org/abs/1701.08251 for more details
+    """
+
+    def __init__(
+        self: AbstractImageTeacher, opt: Dict[str, Any], shared: Dict[str, Any] = None
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: see comment about opt below

you proposed setting a typing helper module. `Shared` is a really good choice for that, would love if you did that here.

File: parlai/tasks/igc/agents.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+from PIL import Image
+from typing import List, Dict, Any
+
+from parlai.core.build_data import download_multiprocess
+from parlai.core.teachers import AbstractImageTeacher
+
+
+class IGCTeacher(AbstractImageTeacher):
+    """
+    Teacher for IGC Task.
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: nit: can you expand out the acronym here

File: parlai/tasks/igc/agents.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        self.multi_ref = opt.get('igc_multi_ref', False)
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Include arg.
+
+        for multi-reference labels.
+        """
+        AbstractImageTeacher.add_cmdline_args(argparser)
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: do `super().add_cmdline_args(argparser)` instead please

File: parlai/tasks/igc/agents.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+            image_id key, for IGC this is a str
+
+        :return:
+            the image path associated with the given image key
+        """
+        if image_id not in self.valid_image_ids:
+            image_id = self.blank_image_id
+        return os.path.join(self.get_image_path(self.opt), image_id)
+
+    def get_data_path(self, opt: Dict[str, Any]) -> str:
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: another opt

Comment (stephenroller) [reply parlai/tasks/igc/agents.py]: also, should this one be private?

File: parlai/tasks/igc/agents.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        Override so that subclasses can see same image features.
+        """
+        # In default implementation, self.data_path already has task name added
+        image_features_path = os.path.join(self.data_path, 'image_features')
+
+        if not os.path.isdir(image_features_path):
+            os.makedirs(image_features_path)
+
+        return os.path.join(
+            image_features_path, '%s_%s_features_dict' % (image_model_name, dt)
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: nit, fstring might look be cool

File: parlai/tasks/igc/agents.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+                if i == 0:
+                    fields = row
+                else:
+                    ep = dict(zip(fields, row))
+                    ep['image_id'] = f'{ep["id"]}'
+                    self.data.append(ep)
+
+        if dt == 'train':
+            # Take first 90% of valid set as train
+            self.data = self.data[: int(len(self.data) * 0.9)]
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: is the data pre-shuffled?

File: parlai/tasks/igc/agents.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+
+class IGCOneSideTeacher(ABC, IGCTeacher):
+    """
+    Override to only return one side of the conversation.
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        IGCTeacher.add_cmdline_args(argparser)
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: similarly, super()

File: parlai/tasks/igc/agents.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+                except OSError:
+                    os.remove(img_path)
+
+    def share(self) -> Dict[str, Any]:
+        shared = super().share()
+        shared['valid_image_ids'] = self.valid_image_ids
+        return shared
+
+
+class IGCOneSideTeacher(ABC, IGCTeacher):
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: why ABC here?

File: parlai/tasks/igc/agents.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        IGCTeacher.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('IGCResponseOnly Arguments')
+        agent.add_argument(
+            '--igc-multi-ref',
+            type='bool',
+            default=False,
+            help='specify true to evaluate on multi-reference labels',
+        )
+
+    def num_episodes(self) -> int:
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: so many types, i'm so happy.

File: tests/tasks/test_igc.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
```
Comment (stephenroller) [inline tests/tasks/test_igc.py]: no way!

File: tests/tasks/test_igc.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+
+@testing_utils.skipIfCircleCI
+class TestIGC(unittest.TestCase):
+    """
+    Basic tests for IGC Teacher to ensure it works for every given variant.
+    """
+
+    def run_display_output(
```
Comment (stephenroller) [inline tests/tasks/test_igc.py]: make private?

File: parlai/tasks/igc/agents.py (hunk 6 above)
Comment (klshuster) [reply parlai/tasks/igc/agents.py]: this function, along with several others, are public methods in the parent `AbstractImageTeacher` class - normally i'd say yes, probably should make private, however these methods are encouraged to be overridden, so not sure what the correct approach is

File: parlai/tasks/igc/agents.py (hunk 8 above)
Comment (klshuster) [reply parlai/tasks/igc/agents.py]: to my knowledge no, however there doesn't appear to be any sense of order in the datafile. this is an arbitrary choice anyway, as there is no train set really...

File: parlai/tasks/igc/agents.py (hunk 10 above)
Comment (klshuster) [reply parlai/tasks/igc/agents.py]: i don't want anyone instantiating this teacher directly, as one should instead use a subclass (response only, question only)

**Comment by klshuster** [comment]
I've addressed your comments @stephenroller. I added a prototype `typing.py` file to `parlai/utils/` --> I wasn't sure exactly what sort of format we wanted to go with here, so please let me know what you think

**Comment by klshuster** [comment]
In the typing file, I've adopted a convention where we define a class for each "type", and then define a type for this class. E.g., shared is now defined as `_Shared` with type `TShared`. If this is not desirable, please let me know.

File: parlai/tasks/igc/agents.py [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+
+from abc import ABC, abstractmethod
+from PIL import Image
+from typing import List, Dict, Any
+
+from parlai.core.build_data import download_multiprocess
+from parlai.core.params import Opt
+from parlai.core.teachers import AbstractImageTeacher
+
+import parlai.utils.typing as PT
```
Comment (stephenroller) [inline parlai/tasks/igc/agents.py]: I would mind mirroring the regular typing. `From parlai.utils.typing import TShared`, but this is cool too. I'm happy.

File: parlai/utils/typing.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Definitions of general ParlAI types.
+"""
+from typing import TypeVar
+
+
+class _Shared(dict):
```
Comment (stephenroller) [inline parlai/utils/typing.py]: huh, I thought somehow it would be like `TShared(Dict[str, Any])`?

**Comment by stephenroller** [comment]
note the current output of mypy:

```
parlai/tasks/igc/agents.py:42: error: Argument 2 for "super" not an instance of argument 1
parlai/tasks/igc/agents.py:46: error: "AbstractImageTeacher" has no attribute "blank_image_id"
Found 2 errors in 1 file (checked 786 source files)
```

**Comment by klshuster** [comment]
it's not immediately clear to me how to fix those so i am just removing the `self: AbstractImageTeacher` typing from init for now...

**Pull request merged**